journal/src/main/java/io/camunda/zeebe/journal/file/MappedJournalSegmentReader.java

```diff
@@ -43,11 +44,18 @@
   }
 
   public boolean hasNext() {
+    checkSegmentOpen();
     // if the next entry exists the version would be non-zero
     return FrameUtil.hasValidVersion(buffer);
   }
 
+  private void checkSegmentOpen() {
+    Preconditions.checkState(
```

Comment:
Here, I'm not sure what would be a good semantics. Ideally readers should reset to a valid position, when the segment is deleted. But due to asynchronous nature, it can happen that the reader is reset only after a delay and try to read from the deleted segment. By failing this operation (probably with an explicit exception), reader can handle this case gracefully.

Reply:
Just so you don't think I'm ignoring this :sweat_smile:, I'm deferring answering this until I understand why we need to guarantee this condition in the `MappedJournalSegmentReader`

Reply:
I'm not fully convinced myself that this check is needed. My arguments for having the check is - The segment is already deleted (semantically). But we do not know if the entries in the segment is committed or not. At the time of writing these entries commit index might be lower. Since the journal is reset, the commit index might have advanced than the indices in this segment.  If a committed reader checks, the commit position is higher than the index it want's to read from the deleted segment. Hence it proceeds to read. But it is possible that the actual committed entries at that index is different, because those entries might have been truncated from other nodes. Hence it can read dirty entries, thinking it is a valid entry.

Reply:
Sorry, I'm still confused :sweat_smile: We only mark a segment for deletion once all its readers have been closed. This means that any `MappedJournalSegmentReader` previously pointing to it had its `close` method closed, right? Semantically, I wouldn't expect any one to use the objects after they've been closed. Do we expect this to happen? My understand was that when marking them for deletion, we would also remove them from the segment map, so that even if the commit index is increased, no `SegmentedJournalReader` would select a segment marked for deletion. Is this the case we want to guard against? Concurrently marking a segment for deletion while another reader is opened for it?

Reply:
> We only mark a segment for deletion once all its readers have been closed. 

No. We mark the segment for deletion when the segment is expected to be deleted. At this time there will be readers opened. Hence we don't delete the segment immediately. But only mark them for deletion, and delete it only once all the readers are closed. 

> My understand was that when marking them for deletion, we would also remove them from the segment map, so that even if the commit index is increased, no SegmentedJournalReader would select a segment marked for deletion.

The segment will be removed from the map. Hence no new readers can be created for this segment. But if a reader is already on the segment, `next()` will read from the segment without this guard.

It can happen that the segment is marked for deletion (meaning that logically it is deleted, but internally it is not garbage collected). Previously it was immediately deleted, resulting in segmentation fault if the reader tries to read from it. By delaying the deletion, we prevent segmentation fault. But the reader will still read from the segment.

Reply:
Ah, gotcha. Then it's definitely required, but I agree that the semantics are strange. So right now, the segment is marked for deletion from the writer context, and any reader pointing to it are unaware of this. We would want them, on the next access, to get reset to the right position, correct?

However, right now, what will happen is the `IllegalStateException` will bubble up all the way to the original caller, e.g. whoever called on `LogStreamReader#hasNext()` - I would expect `hasNext()` to return a boolean, not to throw exceptions, to be honest, and this isn't really part of the contract right now so I expect consumers in our application don't handle it gracefully. If we switch to return false if the segment is closed instead, then it the `SegmentedJournalReader` should try to get from the next segment and reset appropriately, all of this transparently to the consumer. I'm having a hard time wrapping my head around that though - is this really the best approach? Wouldn't the original caller be the one to know if it wants to be reset or do something else? I guess previously it was simply transparently resetting the reader, so we could keep this behavior?

Reply:
> We would want them, on the next access, to get reset to the right position, correct?

At the least, we don't want readers to read dirty values.

Returning false on hasNext() would be a good alternative. But I'm reluctant to reset the readers transparently. Consider this case - StreamProcessor reads position 5, journal was reset to position 15, reader was automatically reset to 15, StreamProcessor reads next event at position 15, StreamProcessor is now in an invalid state. We know the StreamProcessor will be closed and reset with new snapshot and the log. But in between it is in an invalid state.

Reply:
May be we can do the following. If segment is closed, `hasNext()` returns false. It is the responsibility of the application using the journal to reset the readers when the journal is reset. The application is the one resetting the journal. The responsibility of the journal is only to give safe access to the segments and prevent it from reading dirty values. Until the reader is reset, `hasNext()` should return false. Here the reader is `SegmentedJournalReader`, reader for the journal not only for this segment.

Reply:
I was wondering what the previous behavior was, but I guess we have no previous behavior :smile: I assume this is addressed in your draft PR? The way I see it, we have two options:

1. Let the exception bubble up all the way to the original caller (e.g. the stream processor), which should decide on what to do. I think we would need a more specific exception then so you can handle it properly, and it may carry some context that could be useful.
1. Transparently reset the reader, such that you read 5, then next time you read 11. But that implies that on reset, we are fine with "jumping" around, and that also implies that the user of the reader cannot rely on the fact that there are "no gaps" (not sure if that's something we do at such a high level though).
1. Have hasNext return false, which will propagate up the chain, and handle it appropriately at some point by resetting. This has some downsides that you don't know if you need to reset or not - but possibly you don't care? As you mentioned, the stream processor for example will get externally reset, so all we're doing is telling it there is nothing else to read until it gets stopped and replaced, at which point there will be more to read.

The one downside I see with the last one (which I think is what you proposed if I understood correctly), is that it's not obvious from the outside what's going on. It just looks like you reached the end of the log. In this case, it's OK as the consumer is reset, but I do feel it's brittle and I could imagine us running into issues of "why is there nothing more to read there's clearly more things on the log", or having this be hard to extend for other readers. I would still favor this option of the the exception one (since exceptions should be for, well, exceptional cases due to their cost and how they need to be handled) and transparently resetting it (which could lead to some confusion). What do you think?

Reply:
Looks like your option 3 and my proposal is similar. So we are agreeing on something. I also think this is better than the exception. 

> I was wondering what the previous behavior was, but I guess we have no previous behavior 

Previously readers were transparently reset. But there the assumption was that there are no readers -  followers have no stream processors.